### ISSUE TITLE
Request: Create a new function V8Js::isExtensionRegistered($extensionName)

### ISSUE THREAD

**Issue by TysonAndre** [phpv8/v8js#315]
In order to call V8Js::registerExtension(), I currently need to read the contents of a file first. If `isExtensionRegistered` was implemented, I could avoid reading that file, improving performance.

- I could cache that in APCU, but apcu may not be available, or apcu may be cleared
- There may be a lot of files

**Comment by stesie** [comment]
Hey @TysonAndre, I think you have a fair point. Yet I think you'd be better of using snapshots instead of extensions. Code wise (from a JS API point of view) there's next to no difference and you get the bonus that the snapshot is parsed and compiled once (and serialized to binary format) and then just loaded on every `new V8Js(...)` instantiation. Whereas extensions are parsed and recompiled/interpreted each and every time again you intantiate `V8Js`

So you have a great performance boost with using snapshots

... and otherwise extensions aren't superior to the `require` stuff which also parses JS code every time (yet is friendlier with error reporting and debugging)

So please don't get me wrong I'm not making excuses for not implementing it; I'm just curious if there's a valid point I'm overlooking why extensions are (still) a thing. If you weren't aware of snapshots and you might just need more documentation then please also let me know

cheers
~stesie

**Comment by TysonAndre** [comment]
I wasn't aware of snapshots, and the code I'm working on may have predated snapshot support

- Also, https://secure.php.net/manual/en/v8js.construct.php contributed to my lack of awareness of that.
- Storing snapshots in apcu may work, not sure how much memory that would take
  If the snapshot size doubles, does that affect the V8Js startup time at all?

https://github.com/phpv8/v8js/issues/289

---

>  Whereas extensions are parsed and recompiled/interpreted each and every time again you instantiate V8Js

If I understand what you said correctly, are you saying the below (in the latest v8js version, with php7.1)?

1. A long-lived apache worker with pid $PID receives a request (Before this, it didn't use V8Js at all)
2. That request triggers new V8Js(), with an empty list of extensions (at most one instance is expected tob e created per HTTP request)
3. V8Js->registerExtension is called (I think it returns false, because the extension is persistent)
4. It finishes serving this request
5. It gets a new request
6. That request triggers `new V8Js()` again. **At step 6, does invoking the constructor cause v8js to do work loading extensions?**


registerExtension doesn't mention anything that suggests that registering a large number of extensions would impact the performance of `new V8Js()`

```
     * Registers persistent context independent global Javascript extension.
     * NOTE! These extensions exist until PHP is shutdown and they need to be registered before V8 is initialized.
     * For best performance V8 is initialized only once per process thus this call has to be done before any V8Js objects are created!
```

**Comment by TysonAndre** [comment]
Actually, it turns out the main use case of my code is using string building, doesn't call registerExtension, and probably isn't that resource intensive.

- It has i18n in multiple locales, and multiple possible entry points, so I'm not sure if switching would help much (e.g. one snapshot per locale). Haven't tried yet.


registerExtension was used in only one place, and it's infrequently called

**Comment by stesie** [comment]
* yes, if you register a extension V8 processes its JS code on each and every upcoming context creation (which effectively is on every `new V8Js`)
* yes, documentation on php.net site unfortunately is pretty outdated (#289) and doesn't talk about snapshots
* startup performance gets slower with increasing snapshot sizes (simply because there's more stuff to load); yet it in any case is far less then parsing JS code; you might want to have a look at https://stesie.github.io/2016/02/snapshot-performance which compares time needed to start V8 itself with and without snapshots + loading server-side react in it (via parsing and including it in the snapshot)
* putting the snapshot in APCu should work, depending on how often the cache is cleared (due to running full) might be wise to put elsewere though

**Comment by TysonAndre** [comment]
I see. I'll look into that.

**Comment by stesie** [comment]
Won't fix, see discusssion above + extensions now marked deprecated (with pull #330)
